### PR TITLE
Stats: show negative revenue

### LINKED ISSUE #1796: Stats v4: the chart doesn't handle negative revenue
**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.

**To Reproduce**
- Refund some orders on wp-admin so that the WC Admin dashboard shows a negative revenue
- Launch the app

Expected: the chart should show the negative revenue on the chart
Actual: the negative revenue isn't reflected on the chart, and the "No revenue this period" message is shown on top of the chart

**Screenshots**

**Expected behavior**

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: 
- iOS version: 
- WooCommerce iOS version: 

3.4

**WordPress Environment**
<details>
```
Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin.
```
</details>

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #1796 

## Changes

- Allowed negative revenue to show on the stats charts by removing `yAxis.axisMinimum = Constants.chartYAxisMinimum (0.0)`
- Updated y-axis minimum and maximum accessibility values to the min and max of the revenue items
  - `PeriodDataViewController` for stats v3 and `StoreStatsV4PeriodViewController` for stats v4

## Testing

Prerequisites:
- the site has WC version >= 4.0 or has WC Admin plugin installed
- the store has some negative revenue from refunds

### Stats v3

- Launch the app
- Go to Settings > Experimental Features
- Make sure "Improved stats" is turned off
- Go back to the "My store" tab --> the negative revenue should be shown
- Check the accessibility value of the y-axis (with Accessibility Inspector on a simulator or VoiceOver on a device) --> it should contain the correct min and max revenue range

### Stats v4

- Launch the app
- Go to Settings > Experimental Features
- Make sure "Improved stats" is turned on
- Go back to the "My store" tab --> the negative revenue should be shown
- Check the accessibility value of the y-axis (with Accessibility Inspector on a simulator or VoiceOver on a device) --> it should contain the correct min and max revenue range

## Example screenshots

stats v3 | stats v4
-- | --
Screen Shot 2020-02-18 at 9 27 46 AM | Screen Shot 2020-02-18 at 9 32 43 AM

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by pmusolino — CHANGES_REQUESTED** [review]
Thank you Jaclyn for taking care of this issue!

I tested it with Stats v3 and Stats v4, but it seems that it's not working properly with V3 (it doesn't show the negative bar), and it's working good filtering stats by `Today` and `This Week` in Stats V4, but not selecting `This Month` or `This Year`. 🤔 I don't know if it's a missing data issue, or not.

I attach screenshots to better understand the problem. Any idea? 

| V4 Today             |  V4 This Week |  V4 This Month (same for this year) |
:-------------------------:|:-------------------------:|:-------------------------:
Simulator Screen Shot - iPhone 11 Pro - 2020-02-24 at 15 25 53 | Simulator Screen Shot - iPhone 11 Pro - 2020-02-24 at 15 25 55 | v4 after

With Stats V3 seems it doesn't work at all:

| V3 days             |  V3 weeks |  V3 Months (same for this year) |
:-------------------------:|:-------------------------:|:-------------------------:
Simulator Screen Shot - iPhone 11 Pro - 2020-02-24 at 15 32 34 | Simulator Screen Shot - iPhone 11 Pro - 2020-02-24 at 15 32 33 | v4 after

**Comment by shiki** [comment]
Found a VoiceOver issue while testing this https://github.com/woocommerce/woocommerce-ios/issues/1890. I do not think it's caused by this PR though.

File: WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift [hunk 1]
```diff
@@ -368,8 +386,7 @@ extension PeriodDataViewController: IAxisValueFormatter {
                 // Do not show the "0" label on the Y axis
                 return ""
             } else {
-                yAxisMaximum = value.humanReadableString()
-                return CurrencyFormatter().formatCurrency(using: yAxisMaximum,
+                return CurrencyFormatter().formatCurrency(using: value.humanReadableString(),
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift]: 👏 🎉 

I'm glad you removed `yAxisMaximum = value.humanReadableString()`. It was very odd that a `stringForValue(_)` method **mutates** the state (`yAxisMaximum` prop). 😱

File: WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift [hunk 2]
```diff
@@ -38,8 +38,27 @@ class PeriodDataViewController: UIViewController {
     @IBOutlet private weak var chartAccessibilityView: UIView!
 
     private var lastUpdatedDate: Date?
-    private var yAxisMinimum: String = Constants.chartYAxisMinimum.humanReadableString()
-    private var yAxisMaximum: String = ""
+
+    private var revenueItems: [Double] {
+        return orderStats?.items?.map({ $0.grossSales }) ?? []
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift]: It's probably better to move the default value early to give more emphasis on what's important: mapping of `grossSales`. 

```suggestion
        (orderStats?.items ?? []).map(\.grossSales)
```

File: WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift [hunk 3]
```diff
@@ -38,8 +38,27 @@ class PeriodDataViewController: UIViewController {
     @IBOutlet private weak var chartAccessibilityView: UIView!
 
     private var lastUpdatedDate: Date?
-    private var yAxisMinimum: String = Constants.chartYAxisMinimum.humanReadableString()
-    private var yAxisMaximum: String = ""
+
+    private var revenueItems: [Double] {
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift]: Should we name it as `grossSales` instead? I feel like that's easier to understand. 🤔

File: WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift [hunk 4]
```diff
@@ -38,8 +38,27 @@ class PeriodDataViewController: UIViewController {
...
+    }
+
+    private var yAxisMaximum: String {
+        guard let orderStats = orderStats else {
+            return ""
+        }
+        let max = revenueItems.max() ?? 0
+        return CurrencyFormatter().formatHumanReadableAmount(String(max), with: orderStats.currencyCode) ?? ""
+    }
+
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift]: I think we should group these together with the `xAxisMinimum` and `xAxisMaximum` properties.

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift [hunk 5]
```diff
@@ -71,8 +71,29 @@ class StoreStatsV4PeriodViewController: UIViewController {
...
+                                                             with: currencyCode,
+                                                             roundSmallNumbers: false) ?? String()
+    }
+
+    private var yAxisMaximum: String {
+        let max = revenueItems.max() ?? 0
+        return CurrencyFormatter().formatHumanReadableAmount(String(max),
+                                                             with: currencyCode,
+                                                             roundSmallNumbers: false) ?? String()
+    }
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift]: I think we should group these together with the `xAxisMinimum` and `xAxisMaximum` properties.

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift [hunk 6]
```diff
@@ -688,8 +707,7 @@ private extension StoreStatsV4PeriodViewController {
     }
 
     func hasRevenue() -> Bool {
-        let totalRevenue = orderStatsIntervals.map({ $0.revenueValue }).reduce(0, +)
-        return totalRevenue > 0
+        return revenueItems.map({ $0 != 0 }).contains(true)
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift]: We can save a loop by using `contains(where:)`

```suggestion
        revenueItems.contains { $0 != 0 }
```

**Review by shiki — COMMENTED** [review]
Thanks for working on this, @jaclync! 

I'm not seeing the same issues as @pmusolino for v4. 

_ | _ | _
--------|-------|-----
IMG_3260        |       IMG_3259 | IMG_3258

I'm using WC 4.0.

**Comment by pmusolino** [comment]
I forgot to mention that I'm using WC 3.9 with WC Admin.

**Comment by jkmassel** [comment]
I'm freezing `3.7`, so this is being bumped to `3.8`. If you need it to be part of the `3.7` release, please re-target `release/3.7` and merge there. Once you have, please ping me that it's been merged and I'll be happy to cut a new beta release!

**Comment by jaclync** [comment]
> I tested it with Stats v3 and Stats v4, but it seems that it's not working properly with V3 (it doesn't show the negative bar), and it's working good filtering stats by `Today` and `This Week` in Stats V4, but not selecting `This Month` or `This Year`. 🤔 I don't know if it's a missing data issue, or not.

@pmusolino I just tried stats v3 on your store "Blog di wootesting", and had the following charts that look quite different from your screenshots 🤔do you know if I missed something on your site?

A negative value is showing in the "Weeks" tab:

Days | Weeks | Months | Years
-- | -- | -- | --
Simulator Screen Shot - iPhone 11 - 2020-02-25 at 11 34 24 | Simulator Screen Shot - iPhone 11 - 2020-02-25 at 11 34 27 | Simulator Screen Shot - iPhone 11 - 2020-02-25 at 11 34 28 | Simulator Screen Shot - iPhone 11 - 2020-02-25 at 11 34 30

Also, stats v3 uses the WP.com stats endpoint and thus isn't affected by WC Admin. Before WC 4.0, WC Admin is a separate plugin that stores might not have installed, and that's why we still keep stats v3 in case the store doesn't have WC Admin / WC 4.0. We do plan to deprecate stats v3 soon after WC 4.0 is more adopted (p91TBi-2tl-p2)!

**Comment by jaclync** [comment]
@shiki thanks a lot for the suggestions! I updated them in 

> Found a VoiceOver issue while testing this: #1890. I do not think it's caused by this PR though.

that's interesting and thanks for logging it! I tested again on a device, and the VoiceOver was pronouncing the correct negative $ amount for my test store. if possible, could you invite me to your test store?

**Comment by pmusolino** [comment]
Hey @jaclync, I'll invite you to my new testing store. The old is no more active. 😉

**Comment by jaclync** [comment]
thanks @pmusolino for adding me to your new store! I took a look at the v3 charts in your store, and I think the reason the negative bars weren't shown is because the magnitude is too small relative to the revenue of ~7M on Jan 28.

Here's the array of gross sales:
```
▿ 30 elements
  - 0 : 7066830.5
  - 1 : 0.0
  - 2 : 0.0
  - 3 : 0.0
  - 4 : 0.0
  - 5 : 0.0
  - 6 : 0.0
  - 7 : 25.0
  - 8 : 25.0
  - 9 : 0.0
  - 10 : 0.0
  - 11 : 0.0
  - 12 : 0.0
  - 13 : 0.0
  - 14 : 0.0
  - 15 : 0.0
  - 16 : -100.0
  - 17 : 0.0
  - 18 : 0.0
  - 19 : 0.0
  - 20 : 0.0
  - 21 : 0.0
  - 22 : 0.0
  - 23 : 0.0
  - 24 : 293213.32
  - 25 : 0.0
  - 26 : 0.0
  - 27 : -25.0
  - 28 : 0.0
  - 29 : 0.0
```

so there are 2 days with negative sales (e.g. -25.0 on Feb 24 also in stats v4 charts), and I was able to see the callout when tapping on Feb 24 for example:



however, the day with the largest negative revenue of -100.0 is 0.0014% of the largest revenue of 7 million 😅 the chart height would have to be at least 71k px for 0.0014% of it to be 1px

in stats v4, we don't show the day with the 7M revenue, Jan 28, except in the "This Year" chart so we see the -25.0 on Feb 24 in the "This Week" tab 🙂

**Comment by pmusolino** [comment]
@jaclync got it, thanks for explaining and investigating on this 😄  I'd say it's a rather remote case to have over 7 million in revenue and so few refunds.

File: RELEASE-NOTES.txt [hunk 7]
```diff
@@ -1,5 +1,6 @@
 3.8
 -----
+- Dashboard stats: any negative revenue (from refunds for example) for a time period should be shown now.
```
Comment (pmusolino) [inline RELEASE-NOTES.txt]: I'd be inclined to substitute `should` with `will` :D

**Review by pmusolino — APPROVED** [review]
It works well for me 😉 

Code ✅ 
Testing ✅

File: RELEASE-NOTES.txt (hunk 7 above)
Comment (jaclync) [reply RELEASE-NOTES.txt]: good idea! updated here: https://github.com/woocommerce/woocommerce-ios/pull/1874/commits/118fcf571f3d6b8c06871224b7b0f54eee1e9b6a

**Pull request merged**